In [1]:
# Run startup commands
# Note: if cwd is already the companion repo, this will fail.
# Either restart the kernel first or don't run this cell, in that case.
project_dir=$(pdm info --where)
source $project_dir/.bashnbrc

(ins):
(ins):
(ins):
(ins):
(ins):
(ins):
(ins):
(ins):


In [2]:
# Make sure we're in the repo directory
GITHUB_URL='https://github.com/edsq/eeskew-pwg-test-000'  # needed for embed-repo-link
tmp_dir="_tmp_pwg_presentation_02-08-2023_part1"
repo_name="eeskew-pwg-test-000"
cd $project_dir/repos &&
rm -rf $tmp_dir &&
mkdir -p $tmp_dir/$repo_name
git clone $repo_name $tmp_dir/$repo_name &&
cd $tmp_dir/$repo_name

Cloning into '_tmp_pwg_presentation_02-08-2023_part1/eeskew-pwg-test-000'...
Cloning into '_tmp_pwg_presentation_02-08-2023_part1/eeskew-pwg-test-000'...
done.
done.


In [3]:
# Start companion repo fresh from the beginning
git checkout $(git rev-list --topo-order main | tail -1)  # check out first commit
pdm venv create --force python

Note: switching to '52e0b0893775a6d5881fda9d5de8b154d5588ae3'.

Note: switching to '52e0b0893775a6d5881fda9d5de8b154d5588ae3'.
You are in 'detached HEAD' state. You can look around, make experimental

changes and commit them, and you can discard any commits you make in this
You are in 'detached HEAD' state. You can look around, make experimental
state without impacting any branches by switching back to a branch.
changes and commit them, and you can discard any commits you make in this

state without impacting any branches by switching back to a branch.
If you want to create a new branch to retain commits you create, you may

do so (now or later) by using -c with the switch command. Example:
If you want to create a new branch to retain commits you create, you may

do so (now or later) by using -c with the switch command. Example:
  git switch -c <new-branch-name>


  git switch -c <new-branch-name>
Or undo this operation with:


Or undo this operation with:
  git switch -


  git switch

In [4]:
# This cell hidden in presentation and docs
# Check that the environment and project are correct
pdm info

PDM version:
  2.5.3
PDM version:
Python Interpreter:
  2.5.3
  /Users/Ed/python/eds-notes/repos/_tmp_pwg_presentation_02-08-2023_part1/eeskew
Python Interpreter:
-pwg-test-000/.venv/bin/python (3.11)
  /Users/Ed/python/eds-notes/repos/_tmp_pwg_presentation_02-08-2023_part1/eeskew
Project Root:
-pwg-test-000/.venv/bin/python (3.11)
  /Users/Ed/python/eds-notes/repos/_tmp_pwg_presentation_02-08-2023_part1/eeskew
Project Root:
-pwg-test-000
  /Users/Ed/python/eds-notes/repos/_tmp_pwg_presentation_02-08-2023_part1/eeskew
Local Packages:
-pwg-test-000
  
Local Packages:
  


# Project Management with PDM

Part 1 of notes from talk given to the WSU Python Working Group on February 8, 2023.  See Part 2 [here](part2.ipynb).

In this chapter, I show the basics of project management with [PDM](https://pdm.fming.dev/latest/): how to create a project,
add dependencies, and add development dependencies.

This tutorial is primarily aimed at macOS and Linux users, although the commands for Windows should mostly translate.

:::{note}
A companion repository with the example project created in these notes is available [here](https://github.com/edsq/eeskew-pwg-test-000).
:::

## Requirements

- The ability to get python executables of different versions, such as with [pyenv](https://github.com/pyenv/pyenv) or [conda](https://docs.conda.io/en/latest/miniconda.html)
- [PDM](https://pdm.fming.dev/latest/) available globally
- The [pdm-bump](https://github.com/carstencodes/pdm-bump) plugin installed

The version of PDM these notes use is:

In [44]:
pdm --version

PDM, version 2.5.3
PDM, version 2.5.3


## Set the python version and initialize the project

First, create the project directory and `cd` into it:

```
mkdir eeskew-pwg-test-000
cd eeskew-pwg-test-000
```

:::{important}
Because this is a throwaway test project, it is important that you give your project a name that won't conflict with any other package on PyPI or TestPyPI. Adding your name and some numbers is a good way to ensure this.
:::

Here, we'll use python version 3.11, but you may change this to be whatever you like.  I'll cover two methods of setting the python version: using [`pyenv`](https://github.com/pyenv/pyenv), and using [`conda`](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html).

### Using `pyenv`

Install python 3.11 if it is not already (see installed versions with `pyenv versions`):

```
pyenv install 3.11
```

Set the local python version for this project and initialize using that version:

```
pyenv local 3.11
pdm init --python python
```

`pyenv local` creates a file `.python-version`, which `pyenv` reads and redirects the command `python` to the installed `python3.11`.  Thus, we only need to tell pdm to use the usual `python` executable.

### Using `conda`

PDM can also use `conda` to create your virtual environment.  To do this simply, we create the virtual environment before initializing the project, so that we can pass the right python executable to `pdm init`.

```
pdm venv create -w conda 3.11
pdm init --python .venv/bin/python
```

::::{tab-set}
:::{tab-item} Using `pyenv`

Install python 3.11 if it is not already (see installed versions with `pyenv versions`):

```
pyenv install 3.11
```

Set the local python version for this project and initialize using that version:

```
pyenv local 3.11
pdm init --python python
```

`pyenv local` creates a file `.python-version`, which `pyenv` reads and redirects the command `python` to the installed `python3.11`.  Thus, we only need to tell pdm to use the usual `python` executable.
:::

:::{tab-item} Using `conda`

PDM can use `conda` to create your virtual environment.  To do this simply, we create the virtual environment before initializing the project, so that we can pass the right python executable to `pdm init`.

```
pdm venv create -w conda 3.11
pdm init --python .venv/bin/python
```
:::
::::

### `pdm init` options

For

```
Would you like to create a virtualenv with <path-to-python>? [y/n] (y):
```

Ensure you select `y`.  Otherwise, PDM will operate in "PEP 582 mode" - see note on this below.

For

```
Is the project a library that is installable?
If yes, we will need to ask a few more questions to include the project name and build backend [y/n] (n):
```

select `y`.  Otherwise, all the default options should be good, except where you want to fill in your own information (project description, email, etc).

## The PDM project

Let's take a look at what we've created:

In [5]:
ls -a

.   .git        .pdm-python      .venv      pyproject.toml
..  .gitignore  .python-version  README.md
.   .git        .pdm-python      .venv      pyproject.toml
..  .gitignore  .python-version  README.md


The relevant files created are the `README.md`; `.pdm.toml`, which holds local configuration for this PDM project; and `pyproject.toml`, which holds project tool configuration and package metadata.

:::{seealso}
PDM has also created a virtual environment for us in the `.venv` directory.  This is where our package and its dependencies will be installed.  If you are unfamiliar with virtual environments, I recommend [this article](https://realpython.com/python-virtual-environments-a-primer/).
:::

:::{note}
PDM supports the rejected Python Enhancement Proposal (PEP) [582](https://peps.python.org/pep-0582/).  An alternative to virtual environments, PEP 582 would automatically get dependencies from a `__pypackages__` directory in the project root, without having to activate a virtualenv.  See the PDM docs on PEP 582 [here](https://pdm.fming.dev/latest/usage/pep582/).  We will not be using PEP 582 mode in this tutorial.
:::

### The `pyproject.toml` file

In [6]:
embed-repo-link "The central location of our project configuration is {pyproject.toml}":

The central location of our project configuration is [`pyproject.toml`](https://github.com/edsq/eeskew-pwg-test-000/tree/52e0b0893775a6d5881fda9d5de8b154d5588ae3/pyproject.toml):


The central location of our project configuration is [`pyproject.toml`](https://github.com/edsq/eeskew-pwg-test-000/tree/52e0b0893775a6d5881fda9d5de8b154d5588ae3/pyproject.toml):


In [7]:
simple-diff $EMPTY_TREE pyproject.toml | show-code toml

```toml
# pyproject.toml
[tool.pdm]

[project]
name = "eeskew-pwg-test-000"
version = "0.1.0"
description = "A test project for presentation to the WSU Python Working Group."
authors = [
    {name = "Edward Eskew", email = "edward.eskew@wsu.edu"},
]
dependencies = []
requires-python = ">=3.11"
readme = "README.md"
license = {text = "MIT"}

[build-system]
requires = ["pdm-pep517>=1.0"]
build-backend = "pdm.pep517.api"
```


```toml
# pyproject.toml
[tool.pdm]

[project]
name = "eeskew-pwg-test-000"
version = "0.1.0"
description = "A test project for presentation to the WSU Python Working Group."
authors = [
    {name = "Edward Eskew", email = "edward.eskew@wsu.edu"},
]
dependencies = []
requires-python = ">=3.11"
readme = "README.md"
license = {text = "MIT"}

[build-system]
requires = ["pdm-pep517>=1.0"]
build-backend = "pdm.pep517.api"
```


This file is written in `.toml` format, which stands for [Tom's Obvious Minimal Language](https://toml.io/en/).

The `tool.pdm` table is empty, although we'll add things here later on.

The `project` table contains the metadata needed to install our project.  Its values thus far were set by the options we chose while running `pdm init`.

The `build-system` section tells the build frontend (e.g. `pip`) what build backend to use - the build backend is what will actually create the distribution artifacts (wheels and sdists), which we'll see later.  See [PEP 517](https://peps.python.org/pep-0517/) for more information.  Here, we're just using the default PDM backend.

:::{seealso}
See the [PDM docs on writing `pyproject.toml`](https://pdm.fming.dev/latest/reference/pep621/) for more on what can be specified in this file.
:::

## Adding code

First we create our package directory in [src layout](https://hynek.me/articles/testing-packaging/), with an empty (for now) `__init__.py` file to indicate that it is a python package:

In [8]:
mkdir src
mkdir src/eeskew_pwg_test_000
touch src/eeskew_pwg_test_000/__init__.py

In [9]:
# checkpoint
git-checkpoint

Previous HEAD position was 52e0b08 pdm init skeleton
Previous HEAD position was 52e0b08 pdm init skeleton
HEAD is now at 681a2ff Add __init__.py
HEAD is now at 681a2ff Add __init__.py


:::{note}
By convention, we use [snake_case](https://en.wikipedia.org/wiki/Snake_case) for the package name, while we use [kebab-case](https://en.wiktionary.org/wiki/kebab_case) for the repository name.
:::

### Add a module

In [10]:
# This cell hidden in presentation and docs
cat << "EOF" > src/eeskew_pwg_test_000/sarcasm.py
def sarcasm(s):
    """Convert string `s` to sArCaSm TeXt."""
    out = ""
    for i, c in enumerate(s):
        if i % 2 == 0:
            out += c.lower()

        else:
            out += c.upper()

    return out
EOF

In [11]:
# checkpoint
git-checkpoint

Previous HEAD position was 681a2ff Add __init__.py
HEAD is now at fe7dbab Add sarcasm.py
Previous HEAD position was 681a2ff Add __init__.py
HEAD is now at fe7dbab Add sarcasm.py


In [12]:
embed-repo-link "Let's add some code in {src/eeskew_pwg_test_000/sarcasm.py}:"

Let's add some code in [`src/eeskew_pwg_test_000/sarcasm.py`](https://github.com/edsq/eeskew-pwg-test-000/tree/fe7dbabe72819019cb909a4e2656c465009b2d01/src/eeskew_pwg_test_000/sarcasm.py):


Let's add some code in [`src/eeskew_pwg_test_000/sarcasm.py`](https://github.com/edsq/eeskew-pwg-test-000/tree/fe7dbabe72819019cb909a4e2656c465009b2d01/src/eeskew_pwg_test_000/sarcasm.py):


In [13]:
simple-diff HEAD~ src/eeskew_pwg_test_000/sarcasm.py | show-code python

```python
# src/eeskew_pwg_test_000/sarcasm.py
def sarcasm(s):
    """Convert string `s` to sArCaSm TeXt."""
    out = ""
    for i, c in enumerate(s):
        if i % 2 == 0:
            out += c.lower()

        else:
            out += c.upper()

    return out
```


```python
# src/eeskew_pwg_test_000/sarcasm.py
def sarcasm(s):
    """Convert string `s` to sArCaSm TeXt."""
    out = ""
    for i, c in enumerate(s):
        if i % 2 == 0:
            out += c.lower()

        else:
            out += c.upper()

    return out
```


The actual content of this code is not too important for the purposes of these notes, but for completeness, all it does is capitalize and lowercase alternating letters in a string.

## Install the project

To make our code available in the virtual environment, we have to install it:

In [14]:
pdm install

Lock file does not exist
Updating the lock file...
Lock file does not exist
Updating the lock file...
⠋ Fetching hashes for resolved packages...
⠋ Fetching hashes for resolved packages...
🔒 Lock successful
Changes are written to pdm.lock.
🔒 Lock successful
Changes are written to pdm.lock.
⠋ Fetching hashes for resolved packages...
⠋ Fetching hashes for resolved packages...
All packages are synced to date, nothing to do.
All packages are synced to date, nothing to do.
Installing the project as an editable package...
Installing the project as an editable package...
  ✔ Install eeskew-pwg-test-000 0.1.0 successful
  ✔ Install eeskew-pwg-test-000 0.1.0 successful

🎉 All complete!talling eeskew-pwg-test-000 0.1.0...

🎉 All complete!talling eeskew-pwg-test-000 0.1.0...

36m⠋ Installing eeskew-pwg-test-000 0.1.0...




In [15]:
# checkpoint
git-checkpoint

Previous HEAD position was fe7dbab Add sarcasm.py
HEAD is now at 27534a6 Add pdm.lock after first pdm install
Previous HEAD position was fe7dbab Add sarcasm.py
HEAD is now at 27534a6 Add pdm.lock after first pdm install


Now we can import our package:

In [16]:
pdm run python -c 'from eeskew_pwg_test_000.sarcasm import sarcasm; print(sarcasm("Hello world!"))'

hElLo wOrLd!
hElLo wOrLd!


Note we have to type `pdm run` before our command for it to be run within our project environment.

:::{tip}
If you don't want to type `pdm run` every time before a command to be run in the project virtual environment, you can *activate* the environment, which will modify your `sys.prefix` to point to the `.venv` directory.  See the python `venv` docs [here](https://docs.python.org/3/library/venv.html#how-venvs-work) for more on how virtual environments work.

`pdm` will print the command to activate the project virtual environment with the command `pdm venv activate`.  You can copy and paste that output, or, if you want to activate the environment in one line, use:

```bash
eval $(pdm venv activate)
```

To simplify things further, add this as an alias to your `~/.bashrc` or `~/.bash_profile` (and don't forget to restart your shell or `source ~/.bashrc` after):

```bash
# ~/.bashrc
alias pdm-activate='eval $(pdm venv activate)'
```

This will let you activate the environment with the command `pdm-activate`.

You can deactivate an active virtual environment with the command `deactivate`.  See also the [PDM docs on virtualenv activation](https://pdm.fming.dev/latest/usage/venv/#activate-a-virtualenv).
:::

### The `pdm.lock` file

Running `pdm install` also created a new file, `pdm.lock`:

In [17]:
ls

README.md  pdm.lock  pyproject.toml  src
README.md  pdm.lock  pyproject.toml  src


In [18]:
simple-diff HEAD~ pdm.lock | show-code toml

```toml
# pdm.lock
# This file is @generated by PDM.
# It is not intended for manual editing.

[metadata]
lock_version = "4.2"
groups = ["default"]
content_hash = "sha256:86165c41f17b4b263a688544a3ebc55eccc1713dd177c40649b2e936dab66751"

[metadata.files]
```


```toml
# pdm.lock
# This file is @generated by PDM.
# It is not intended for manual editing.

[metadata]
lock_version = "4.2"
groups = ["default"]
content_hash = "sha256:86165c41f17b4b263a688544a3ebc55eccc1713dd177c40649b2e936dab66751"

[metadata.files]
```


This is a lockfile, which will contain the exact versions of each project dependency we install.  It is useful for creating a perfect reproduction of the project virtual environment, which keeps our development reproducible over time and across different machines.

Right now, we have not installed anything other than the project itself, so it is essentially empty.

## Add a dependency

Let's add a dependency to our project, [cowsay](https://pypi.org/project/cowsay/):

In [19]:
pdm add cowsay

Adding packages to default dependencies: cowsay
Adding packages to default dependencies: cowsay
⠸ Fetching hashes for resolved packages...
⠸ Fetching hashes for resolved packages...
🔒 Lock successful
🔒 Lock successful
Changes are written to pyproject.toml.
Changes are written to pyproject.toml.
⠋ Fetching hashes for resolved packages...
⠋ Fetching hashes for resolved packages...
Synchronizing working set with lock file: 1 to add, 0 to update, 0 to remove

Synchronizing working set with lock file: 1 to add, 0 to update, 0 to remove

  ✔ Install cowsay 5.0 successful
  ✔ Install cowsay 5.0 successful
Installing the project as an editable package...
Installing the project as an editable package...0m...
  ✔ Update eeskew-pwg-test-000 0.1.0 -> 0.1.0 successful
  ✔ Update eeskew-pwg-test-000 0.1.0 -> 0.1.0 successful

🎉 All complete!ating eeskew-pwg-test-000 0.1.0 -> 0.1.0...

🎉 All complete!ating eeskew-pwg-test-000 0.1.0 -> 0.1.0...

36m⠙ Updating eeskew-pwg-test-000 0.1.0 -> 0.1.0...




In [20]:
# checkpoint
git-checkpoint

Previous HEAD position was 27534a6 Add pdm.lock after first pdm install
Previous HEAD position was 27534a6 Add pdm.lock after first pdm install
HEAD is now at fcc0838 Add cowsay to dependencies
HEAD is now at fcc0838 Add cowsay to dependencies


### What did `pdm add` do?

In [21]:
embed-repo-link "\`cowsay\` now appears as a dependency in {pyproject.toml}:"

`cowsay` now appears as a dependency in [`pyproject.toml`](https://github.com/edsq/eeskew-pwg-test-000/tree/fcc083809724f6472f436beb22d76abfca56342c/pyproject.toml):


`cowsay` now appears as a dependency in [`pyproject.toml`](https://github.com/edsq/eeskew-pwg-test-000/tree/fcc083809724f6472f436beb22d76abfca56342c/pyproject.toml):


In [22]:
simple-diff HEAD~ pyproject.toml | show-code toml

```toml
# pyproject.toml
...

authors = [
    {name = "Edward Eskew", email = "edward.eskew@wsu.edu"},
]
dependencies = [
    "cowsay>=5.0",
]
requires-python = ">=3.11"
readme = "README.md"
license = {text = "MIT"}

...
```


```toml
# pyproject.toml
...

authors = [
    {name = "Edward Eskew", email = "edward.eskew@wsu.edu"},
]
dependencies = [
    "cowsay>=5.0",
]
requires-python = ">=3.11"
readme = "README.md"
license = {text = "MIT"}

...
```


In [23]:
embed-repo-link "We've also updated {pdm.lock} to include cowsay:"

We've also updated [`pdm.lock`](https://github.com/edsq/eeskew-pwg-test-000/tree/fcc083809724f6472f436beb22d76abfca56342c/pdm.lock) to include cowsay:


We've also updated [`pdm.lock`](https://github.com/edsq/eeskew-pwg-test-000/tree/fcc083809724f6472f436beb22d76abfca56342c/pdm.lock) to include cowsay:


In [24]:
simple-diff HEAD~ pdm.lock | show-code toml

```toml
# pdm.lock
# This file is @generated by PDM.
# It is not intended for manual editing.

[[package]]
name = "cowsay"
version = "5.0"
summary = "The famous cowsay for GNU/Linux is now available for python"

[metadata]
lock_version = "4.2"
groups = ["default"]
content_hash = "sha256:a04c8eb9409090bc9acb94e5cec5ef19afdecacb2c169628142af5ca472135f0"

[metadata.files]
"cowsay 5.0" = [
    {url = "https://files.pythonhosted.org/packages/6b/b8/9f497fd045d74fe21d91cbe8debae0b451229989e35b539d218547d79fc6/cowsay-5.0.tar.gz", hash = "sha256:c00e02444f5bc7332826686bd44d963caabbaba9a804a63153822edce62bbbf3"},
]
```


```toml
# pdm.lock
# This file is @generated by PDM.
# It is not intended for manual editing.

[[package]]
name = "cowsay"
version = "5.0"
summary = "The famous cowsay for GNU/Linux is now available for python"

[metadata]
lock_version = "4.2"
groups = ["default"]
content_hash = "sha256:a04c8eb9409090bc9acb94e5cec5ef19afdecacb2c169628142af5ca472135f0"

[metadata.files]
"cowsay 5.0" = [
    {url = "https://files.pythonhosted.org/packages/6b/b8/9f497fd045d74fe21d91cbe8debae0b451229989e35b539d218547d79fc6/cowsay-5.0.tar.gz", hash = "sha256:c00e02444f5bc7332826686bd44d963caabbaba9a804a63153822edce62bbbf3"},
]
```


We can now import `cowsay`:

In [25]:
pdm run python -c 'import cowsay; cowsay.cow("moo!")'

  ____
| moo! |
  ____
  ====
| moo! |
    \
  ====
     \
    \
       ^__^
     \
       (oo)\_______
       ^__^
       (__)\       )\/\
       (oo)\_______
           ||----w |
       (__)\       )\/\
           ||     ||
           ||----w |
           ||     ||


:::{seealso}
See the [PDM docs on managing dependencies](https://pdm.fming.dev/latest/usage/dependency/) for more information.
:::

### Adding more code

In [26]:
# This cell hidden in presentation and docs
cat << "EOF" > src/eeskew_pwg_test_000/sarcasm.py
import cowsay

def sarcasm(s):
    """Convert string `s` to sArCaSm TeXt."""
    out = ""
    for i, c in enumerate(s):
        if i % 2 == 0:
            out += c.lower()

        else:
            out += c.upper()

    return out

def sarcastic_cowsay(s):
    """Cowsay `s`, sArCaStIcAlLy."""
    sarcastic_s = sarcasm(s)
    cowsay.cow(sarcastic_s)
EOF

In [27]:
# checkpoint
git-checkpoint

Previous HEAD position was fcc0838 Add cowsay to dependencies
Previous HEAD position was fcc0838 Add cowsay to dependencies
HEAD is now at f9000d5 Add sarcastic_cowsay to utils
HEAD is now at f9000d5 Add sarcastic_cowsay to utils


In [28]:
embed-repo-link "Let's add a new function to {src/eeskew_pwg_test_000/sarcasm.py}:"

Let's add a new function to [`src/eeskew_pwg_test_000/sarcasm.py`](https://github.com/edsq/eeskew-pwg-test-000/tree/f9000d593a66d666701d20e94845c47ca8453a73/src/eeskew_pwg_test_000/sarcasm.py):


Let's add a new function to [`src/eeskew_pwg_test_000/sarcasm.py`](https://github.com/edsq/eeskew-pwg-test-000/tree/f9000d593a66d666701d20e94845c47ca8453a73/src/eeskew_pwg_test_000/sarcasm.py):


In [29]:
simple-diff --context 0 HEAD~ src/eeskew_pwg_test_000/sarcasm.py | show-code python

```python
# src/eeskew_pwg_test_000/sarcasm.py
import cowsay


...


def sarcastic_cowsay(s):
    """Cowsay `s`, sArCaStIcAlLy."""
    sarcastic_s = sarcasm(s)
    cowsay.cow(sarcastic_s)
```


```python
# src/eeskew_pwg_test_000/sarcasm.py
import cowsay


...


def sarcastic_cowsay(s):
    """Cowsay `s`, sArCaStIcAlLy."""
    sarcastic_s = sarcasm(s)
    cowsay.cow(sarcastic_s)
```


We can now run this new function:

In [30]:
pdm run python -c 'from eeskew_pwg_test_000.sarcasm import sarcastic_cowsay; sarcastic_cowsay("mooo!")'

  _____
| mOoO! |
  _____
  =====
| mOoO! |
     \
  =====
      \
     \
        ^__^
      \
        (oo)\_______
        ^__^
        (__)\       )\/\
        (oo)\_______
            ||----w |
        (__)\       )\/\
            ||     ||
            ||----w |
            ||     ||


:::{note}
We didn't have to re-run `pdm install` to use our new function - this is because PDM installs our `eeskew_pwg_test_000` package in "editable mode", which acts sort of like a symlink between the source code and the installed files in the `.venv` directory.
:::

## Add a development dependency

The dependencies listed in the `project.dependencies` section of `pyproject.toml` will all be installed when someone runs `pip install eeskew-pwg-test-000`.  What if we have dependencies we only want in our development environment?

Let's add `black`, a tool to automatically format our code:

In [31]:
pdm add -d black

Adding group dev to lockfile
Adding group dev to lockfile
Adding packages to dev dev-dependencies: black
Adding packages to dev dev-dependencies: black
⠇ Fetching hashes for resolved packages...m23.3.0
⠇ Fetching hashes for resolved packages...
🔒 Lock successful
🔒 Lock successful
Changes are written to pyproject.toml.
Changes are written to pyproject.toml.
⠋ Fetching hashes for resolved packages...
⠋ Fetching hashes for resolved packages...
Synchronizing working set with lock file: 6 to add, 0 to update, 0 to remove

Synchronizing working set with lock file: 6 to add, 0 to update, 0 to remove

  ⠋ Installing black 23.3.0...
  ⠋ Installing black 23.3.0...
  ⠋ Installing black 23.3.0...         
  ⠋ Installing click 8.1.3...          
  ⠋ Installing black 23.3.0...         
  ⠋ Installing click 8.1.3...          
  ⠋ Installing black 23.3.0...         
  ⠋ Installing click 8.1.3...          
  ⠋ Installing black 23.3.0...         
  ⠋ Installing mypy-extensions 1.0.0...
  ⠋ Installing cl

  ⠙ Installing mypy-extensions 1.0.0...
  ⠙ Installing pathspec 0.11.1...      
  ⠙ Installing packaging 23.1...       
  ⠙ Installing pathspec 0.11.1...      
  ✔ Install black 23.3.0 successful
  ⠙ Installing black 23.3.0...         
  ✔ Install black 23.3.0 successful
  ⠙ Installing click 8.1.3...          
  ⠙ Installing black 23.3.0...         
  ⠙ Installing mypy-extensions 1.0.0...
  ⠙ Installing click 8.1.3...          
  ⠙ Installing packaging 23.1...       
  ⠙ Installing mypy-extensions 1.0.0...
  ⠙ Installing pathspec 0.11.1...      
  ⠙ Installing packaging 23.1...       
  ⠙ Installing pathspec 0.11.1...      

🎉 All complete!talling platformdirs 3.4.0...   

  ⠙ Installing black 23.3.0...         
🎉 All complete!
  ⠙ Installing click 8.1.3...          
  ⠙ Installing black 23.3.0...         
  ⠙ Installing mypy-extensions 1.0.0...
  ⠙ Installing click 8.1.3...          
  ⠙ Installing packaging 23.1...       
  ⠙ Installing mypy-extensions 1.0.0...
  ⠙ Installing pathspe

In [32]:
# checkpoint
git-checkpoint

Previous HEAD position was f9000d5 Add sarcastic_cowsay to utils
Previous HEAD position was f9000d5 Add sarcastic_cowsay to utils
HEAD is now at 69a4552 Add black to dev dependencies
HEAD is now at 69a4552 Add black to dev dependencies


### What did `pdm add -d` do?

In [33]:
embed-repo-link "We've added a new \`[dev-dependencies]\` sub-table to the \`[tool.pdm]\` table of {pyproject.toml}, which contains \`black\`:"

We've added a new `[dev-dependencies]` sub-table to the `[tool.pdm]` table of [`pyproject.toml`](https://github.com/edsq/eeskew-pwg-test-000/tree/69a4552574e7a407d1097c91d9bb3129d9db4e88/pyproject.toml), which contains `black`:


We've added a new `[dev-dependencies]` sub-table to the `[tool.pdm]` table of [`pyproject.toml`](https://github.com/edsq/eeskew-pwg-test-000/tree/69a4552574e7a407d1097c91d9bb3129d9db4e88/pyproject.toml), which contains `black`:


In [34]:
simple-diff --context 2 HEAD~ pyproject.toml | show-code toml

```toml
# pyproject.toml
[tool.pdm]
[tool.pdm.dev-dependencies]
dev = [
    "black>=23.3.0",
]

[project]

...
```


```toml
# pyproject.toml
[tool.pdm]
[tool.pdm.dev-dependencies]
dev = [
    "black>=23.3.0",
]

[project]

...
```


When we run `pdm install`, by default, PDM will install dependencies from here in addition to the dependencies listed in `[project.dependencies]`.  A different tool like `pip`, however, will not.

:::{seealso}
See the [PDM docs on adding development dependencies](https://pdm.fming.dev/latest/usage/dependency/#add-development-only-dependencies) for more information on development dependencies.
:::

The lockfile has also been updated:

In [35]:
cat pdm.lock

# This file is @generated by PDM.
# It is not intended for manual editing.
# This file is @generated by PDM.

# It is not intended for manual editing.
[[package]]

name = "black"
[[package]]
version = "23.3.0"
name = "black"
requires_python = ">=3.7"
version = "23.3.0"
summary = "The uncompromising code formatter."
requires_python = ">=3.7"
dependencies = [
summary = "The uncompromising code formatter."
    "click>=8.0.0",
dependencies = [
    "mypy-extensions>=0.4.3",
    "click>=8.0.0",
    "packaging>=22.0",
    "mypy-extensions>=0.4.3",
    "pathspec>=0.9.0",
    "packaging>=22.0",
    "platformdirs>=2",
    "pathspec>=0.9.0",
]
    "platformdirs>=2",

]
[[package]]

name = "click"
[[package]]
version = "8.1.3"
name = "click"
requires_python = ">=3.7"
version = "8.1.3"
summary = "Composable command line interface toolkit"
requires_python = ">=3.7"
dependencies = [
summary = "Composable command line interface toolkit"
    "colorama; platform_system == \"Windows\"",
dependencies = [


    {url = "https://files.pythonhosted.org/packages/3f/0d/81dd4194ce7057c199d4f28e4c2a885082d9d929e7a55c514b23784f7787/black-23.3.0-cp311-cp311-win_amd64.whl", hash = "sha256:11c410f71b876f961d1de77b9699ad19f939094c3a677323f43d7a29855fe326"},
    {url = "https://files.pythonhosted.org/packages/49/d7/f3b7da6c772800f5375aeb050a3dcf682f0bbeb41d313c9c2820d0156e4e/black-23.3.0-cp39-cp39-macosx_10_16_x86_64.whl", hash = "sha256:92c543f6854c28a3c7f39f4d9b7694f9a6eb9d3c5e2ece488c327b6e7ea9b266"},
    {url = "https://files.pythonhosted.org/packages/49/36/15d2122f90ff1cd70f06892ebda777b650218cf84b56b5916a993dc1359a/black-23.3.0-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl", hash = "sha256:50cb33cac881766a5cd9913e10ff75b1e8eb71babf4c7104f2e9c52da1fb7de2"},
    {url = "https://files.pythonhosted.org/packages/69/49/7e1f0cf585b0d607aad3f971f95982cc4208fc77f92363d632d23021ee57/black-23.3.0-cp311-cp311-macosx_10_16_x86_64.whl", hash = "sha256:a6f6886c9869d4daae2d1715ce34a19bbc4b95006d20ed

]
    {url = "https://files.pythonhosted.org/packages/c2/f1/df59e28c642d583f7dacffb1e0965d0e00b218e0186d7858ac5233dce840/click-8.1.3-py3-none-any.whl", hash = "sha256:bb4d8133cb15a609f44e8213d9b391b0809795062913b383c62be0ee95b1db48"},
"colorama 0.4.6" = [
]
    {url = "https://files.pythonhosted.org/packages/d1/d6/3965ed04c63042e047cb6a3e6ed1a63a35087b6a609aa3a15ed8ac56c221/colorama-0.4.6-py2.py3-none-any.whl", hash = "sha256:4f1d9991f5acc0ca119f9d443620b77f9d6b33703e51011c16baf57afb285fc6"},
"colorama 0.4.6" = [
    {url = "https://files.pythonhosted.org/packages/d8/53/6f443c9a4a8358a93a6792e2acffb9d9d5cb0a5cfd8802644b7b1c9a02e4/colorama-0.4.6.tar.gz", hash = "sha256:08695f5cb7ed6e0531a20572697297273c47b8cae5a63ffc6d6ed5c201be6e44"},
    {url = "https://files.pythonhosted.org/packages/d1/d6/3965ed04c63042e047cb6a3e6ed1a63a35087b6a609aa3a15ed8ac56c221/colorama-0.4.6-py2.py3-none-any.whl", hash = "sha256:4f1d9991f5acc0ca119f9d443620b77f9d6b33703e51011c16baf57afb285fc6"},
]
    {url = "h

Many new packages now exist in the lockfile, not just `cowsay` and `black`.  This is because (unlike `cowsay`), `black` itself has dependencies that we needed to install to get it to work.  This lock file records exactly the versions of those sub-dependencies that we've now installed into our project virtual environment.

When we run `pdm install`, if the lock file exists (and `pyproject.toml` hasn't been changed since the lockfile was last updated), PDM will install precisely the packages listed in the lockfile, so we'll always be working in the same virtual environment.  This is useful for developing and testing the code, so you should always include the lockfile in your project version control.

However, we don't want to impose these restrictions on users of our library, or our project would rapidly become impossible to install due to other packages requiring different versions of the packages in the lockfile.  The only thing that we care about is that users have the right versions of the dependencies we directly use, which are listed in the `project.dependencies` array in `pyproject.toml`.  This is why `pip install` does not care about the existence of the lockfile.

:::{seealso}
See the [PDM docs on version control](https://pdm.fming.dev/latest/usage/project/#working-with-version-control) for more on best practices for version-controlling a PDM project.
:::

### Using black

We can now run `black` within our environment.

In [36]:
# This cell hidden in presentation and docs
cat << "EOF" > src/eeskew_pwg_test_000/sarcasm.py
import cowsay
def sarcasm(
            s
          ):
    """Convert string `s` to sArCaSm TeXt."""
    out = ''
    for i,c in \
        enumerate( s ):

        if i% 2 ==0: out +=c.lower()

        else:



            out+= c.upper()

    return out

def sarcastic_cowsay(s):
    """Cowsay `s`, sArCaStIcAlLy."""
    sarcastic_s = sarcasm(s); cowsay.cow(sarcastic_s)
EOF

In [37]:
# checkpoint
git-checkpoint

Previous HEAD position was 69a4552 Add black to dev dependencies
Previous HEAD position was 69a4552 Add black to dev dependencies
HEAD is now at 5c8fbf3 Use bad formatting in utils for black to fix
HEAD is now at 5c8fbf3 Use bad formatting in utils for black to fix


In [38]:
embed-repo-link "Let's re-write {src/eeskew_pwg_test_000/sarcasm.py} with deliberately poor formatting:"

Let's re-write [`src/eeskew_pwg_test_000/sarcasm.py`](https://github.com/edsq/eeskew-pwg-test-000/tree/5c8fbf3be8330a7da646bb07a975276bbf5f969c/src/eeskew_pwg_test_000/sarcasm.py) with deliberately poor formatting:


Let's re-write [`src/eeskew_pwg_test_000/sarcasm.py`](https://github.com/edsq/eeskew-pwg-test-000/tree/5c8fbf3be8330a7da646bb07a975276bbf5f969c/src/eeskew_pwg_test_000/sarcasm.py) with deliberately poor formatting:


In [39]:
simple-diff HEAD~ src/eeskew_pwg_test_000/sarcasm.py | show-code python

```python
# src/eeskew_pwg_test_000/sarcasm.py
import cowsay
def sarcasm(
            s
          ):
    """Convert string `s` to sArCaSm TeXt."""
    out = ''
    for i,c in \
        enumerate( s ):

        if i% 2 ==0: out +=c.lower()

        else:



            out+= c.upper()

    return out

def sarcastic_cowsay(s):
    """Cowsay `s`, sArCaStIcAlLy."""
    sarcastic_s = sarcasm(s); cowsay.cow(sarcastic_s)
```


```python
# src/eeskew_pwg_test_000/sarcasm.py
import cowsay
def sarcasm(
            s
          ):
    """Convert string `s` to sArCaSm TeXt."""
    out = ''
    for i,c in \
        enumerate( s ):

        if i% 2 ==0: out +=c.lower()

        else:



            out+= c.upper()

    return out

def sarcastic_cowsay(s):
    """Cowsay `s`, sArCaStIcAlLy."""
    sarcastic_s = sarcasm(s); cowsay.cow(sarcastic_s)
```


This is bad formatting!  Rather than fix it manually, we can run `black` on our code, which will auto-impose a reasonable style.

In [40]:
pdm run black src/

reformatted /Users/Ed/python/eds-notes/repos/_tmp_pwg_presentation_02-08-2023_part1/eeskew-pwg-test-000/src/eeskew_pwg_test_000/sarcasm.py
reformatted /Users/Ed/python/eds-notes/repos/_tmp_pwg_presentation_02-08-2023_part1/eeskew-pwg-test-000/src/eeskew_pwg_test_000/sarcasm.py

All done! ✨ 🍰 ✨

1 file reformatted, 1 file left unchanged.
All done! ✨ 🍰 ✨
1 file reformatted, 1 file left unchanged.


In [41]:
# checkpoint
git-checkpoint

Previous HEAD position was 5c8fbf3 Use bad formatting in utils for black to fix
Previous HEAD position was 5c8fbf3 Use bad formatting in utils for black to fix
HEAD is now at 7ff4bbc Fix formatting with black
HEAD is now at 7ff4bbc Fix formatting with black


### What did `black` do?

In [42]:
embed-repo-link "\`black\` has automatically re-formatted {src/eeskew_pwg_test_000/sarcasm.py}, fixing the poor formatting we introduced earlier:"

`black` has automatically re-formatted [`src/eeskew_pwg_test_000/sarcasm.py`](https://github.com/edsq/eeskew-pwg-test-000/tree/7ff4bbc4d1775441227ff0c34f1e6de3a4b58f61/src/eeskew_pwg_test_000/sarcasm.py), fixing the poor formatting we introduced earlier:


`black` has automatically re-formatted [`src/eeskew_pwg_test_000/sarcasm.py`](https://github.com/edsq/eeskew-pwg-test-000/tree/7ff4bbc4d1775441227ff0c34f1e6de3a4b58f61/src/eeskew_pwg_test_000/sarcasm.py), fixing the poor formatting we introduced earlier:


In [43]:
simple-diff HEAD~ src/eeskew_pwg_test_000/sarcasm.py | show-code python

```python
# src/eeskew_pwg_test_000/sarcasm.py
import cowsay


def sarcasm(s):
    """Convert string `s` to sArCaSm TeXt."""
    out = ""
    for i, c in enumerate(s):
        if i % 2 == 0:
            out += c.lower()

        else:
            out += c.upper()

    return out


def sarcastic_cowsay(s):
    """Cowsay `s`, sArCaStIcAlLy."""
    sarcastic_s = sarcasm(s)
    cowsay.cow(sarcastic_s)
```


```python
# src/eeskew_pwg_test_000/sarcasm.py
import cowsay


def sarcasm(s):
    """Convert string `s` to sArCaSm TeXt."""
    out = ""
    for i, c in enumerate(s):
        if i % 2 == 0:
            out += c.lower()

        else:
            out += c.upper()

    return out


def sarcastic_cowsay(s):
    """Cowsay `s`, sArCaStIcAlLy."""
    sarcastic_s = sarcasm(s)
    cowsay.cow(sarcastic_s)
```


:::{seealso}
See the [black documentation](https://black.readthedocs.io/en/stable/) for more information.
:::

## Conclusion

That's all for Part 1!  The examples here should be enough to get you started using PDM to manage your own python projects.  For more advanced usage, check out the PDM documentation on:

- [General project management](https://pdm.fming.dev/latest/usage/project/)
- [Working with virtualenv](https://pdm.fming.dev/latest/usage/venv/)
- [Managing dependencies](https://pdm.fming.dev/latest/usage/dependency/)

In [Part 2](part2.ipynb), we'll cover how to publish a PDM project on PyPI, so that it can be installed with a simple `pip install`.